src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractExtensionMethodImportCompletionProvider.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract class AbstractExtensionMethodImportCompletionProvider : AbstractImportCompletionProvider
+    {
+        protected abstract string GenericSuffix { get; }
+
+        protected override bool ShouldProvideCompletion(Document document, SyntaxContext syntaxContext)
+            => syntaxContext.IsRightOfNameSeparator && IsAddingImportsSupported(document);
+
+        protected async override Task AddCompletionItemsAsync(
+            CompletionContext completionContext,
+            SyntaxContext syntaxContext,
+            HashSet<string> namespaceInScope,
+            bool isExpandedCompletion,
+            CancellationToken cancellationToken)
+        {
+            using (Logger.LogBlock(FunctionId.Completion_ExtensionMethodImportCompletionProvider_GetCompletionItemsAsync, cancellationToken))
+            {
+                var syntaxFacts = completionContext.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+                if (TryGetReceiverTypeSymbol(syntaxContext, syntaxFacts, out var receiverTypeSymbol))
+                {
+                    var items = await ExtensionMethodImportCompletionHelper.GetUnimportExtensionMethodsAsync(
+                        completionContext.Document,
+                        completionContext.Position,
+                        receiverTypeSymbol,
+                        namespaceInScope,
+                        isExpandedCompletion,
+                        cancellationToken).ConfigureAwait(false);
+
+                    completionContext.AddItems(items.Select(Convert));
```

Comment:
technically allocates.  consider `items.Select(i => Convert(i))`

Reply:
what's the difference?

Reply:
the lambda form can (if it doesn't capture) be converted to a single-cached allocation by the compiler.  The delegate-conversion form cannot be. So it will allocate each time that `items.Select(Convert)` line is hit.